modules/sql-engine/src/test/java/org/apache/ignite/internal/sql/engine/prepare/TypeCoercionTest.java

```diff
@@ -0,0 +1,395 @@
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.function.BiConsumer;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+import org.apache.calcite.avatica.util.TimeUnit;
+import org.apache.calcite.rel.type.RelDataType;
+import org.apache.calcite.runtime.CalciteContextException;
+import org.apache.calcite.sql.SqlCall;
+import org.apache.calcite.sql.SqlIntervalQualifier;
+import org.apache.calcite.sql.SqlNode;
+import org.apache.calcite.sql.SqlOperator;
+import org.apache.calcite.sql.SqlSelect;
+import org.apache.calcite.sql.fun.SqlStdOperatorTable;
+import org.apache.calcite.sql.parser.SqlParseException;
+import org.apache.calcite.sql.parser.SqlParserPos;
+import org.apache.calcite.sql.type.SqlTypeName;
+import org.apache.ignite.internal.sql.engine.planner.AbstractPlannerTest;
+import org.apache.ignite.internal.sql.engine.schema.IgniteSchema;
+import org.apache.ignite.internal.sql.engine.trait.IgniteDistributions;
+import org.jetbrains.annotations.Nullable;
+import org.junit.jupiter.api.Assumptions;
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.MethodSource;
+
+/**
+ * Tests type coercion behaviour.
+ *
+ * @see IgniteTypeCoercion
+ */
+public class TypeCoercionTest extends AbstractPlannerTest {
+
+    private static final RelDataType VARCHAR = TYPE_FACTORY.createSqlType(SqlTypeName.VARCHAR, 20);
+
+    private static final List<RelDataType> NUMERIC_TYPES =  SqlTypeName.NUMERIC_TYPES.stream().map(t -> {
+        if (t == SqlTypeName.DECIMAL) {
+            return TYPE_FACTORY.createSqlType(t, 10, 2);
+        } else {
+            return TYPE_FACTORY.createSqlType(t);
+        }
+    }).collect(Collectors.toList());
+
+    @ParameterizedTest
+    @MethodSource("booleanToNumeric")
+    public void testBooleanToNumeric(TypeCoercionRule rule) {
+        Tester tester = new Tester(rule);
+        tester.execute();
+    }
+
+    private static Stream<TypeCoercionRule> booleanToNumeric() {
+        List<TypeCoercionRule> numericRules = new ArrayList<>();
+
+        for (RelDataType type : NUMERIC_TYPES) {
+            RelDataType booleanType = TYPE_FACTORY.createSqlType(SqlTypeName.BOOLEAN);
+            numericRules.add(typeCoercionRule(type, booleanType, new ToSpecificType(type)));
+            numericRules.add(typeCoercionRule(booleanType, type, new ToSpecificType(type)));
+        }
+
+        return numericRules.stream();
+    }
+
+    @ParameterizedTest
+    @MethodSource("varCharToNumeric")
+    public void testVarCharToNumeric(TypeCoercionRule rule) {
+        Tester tester = new Tester(rule);
+        tester.execute();
+    }
+
+    private static Stream<TypeCoercionRule> varCharToNumeric() {
+        List<TypeCoercionRule> numericRules = new ArrayList<>();
+
+        for (RelDataType type : NUMERIC_TYPES) {
+            if (type.getSqlTypeName() == SqlTypeName.DECIMAL) {
+                // The widest decimal possible?
```

Comment:
this one definitely deserve the issue to investigate

Reply:
Done